FASTA redesign

laughedelic opened the pull request:
Fixes #35, Fixes #38 

See https://github.com/ohnosequences/fastarious/pull/24#issuecomment-281395442. Let's do the "plain old classes" thing here too.

laughedelic committed:
Changing FASTA to plain old classes (WIP)

laughedelic committed:
Merge branch 'master' into pr/25

laughedelic committed:
Finished refactoring of the FASTA code

laughedelic committed:
Updated fasta test code

laughedelic commented:
Fixes #35

eparejatobes commented on src/main/scala/fasta.scala:
```diff
@@ -26,107 +25,59 @@ import java.io._
 
-    implicit def fastaOps(fa: FASTA.Value): FASTAOps = new FASTAOps(fa)
-    implicit def fastaIteratorOps(fas: Iterator[FASTA.Value]): FASTAIteratorOps = new FASTAIteratorOps(fas)
+    /* Note that description will keep the initial empty space, so that `value == s"${id}${description}"` */
```
the space is not part of the description. This should return the description without the first space.

laughedelic replied:
OK. That's something I actually didn't change here. But I will fix it.

laughedelic replied:
Can the description have trailing spaces? or multiple spaces between ID and description? If not, I'll just use `.trim`

laughedelic replied:
Fixed in 5bd842d (without `.trim`)

eparejatobes commented on src/main/scala/fasta.scala:
```diff
@@ -145,30 +96,30 @@ case object fasta {
-      def next(): Either[ParseDenotationsError, FASTA.Value] = {
-        val hdr: String = lines.next()
-        val seq: String = cutUntil(fasta.header.is).mkString
+      def next(): Option[FASTA] = {
```
this should call hasNext first

laughedelic replied:
Fixed in c9ff593. But I have another concern about this iterator:
- it should either return `FASTA` without `Option`, because `hasNext` already checks that the next line is a valid header
- or `hasNext` should do that check

So you either have an Iterator, which reads all lines and returns `None` for chunks which are not a valid FASTA, or you have an iterator which reads only consecutive valid `FASTA` lines (and then it doesn't need to return `Option`).

laughedelic replied:
P.S. we have to choose one or another way to modify it, because currently it just won't ever return `None`s.

laughedelic replied:
In my opinion `parseFasta: Iterator[Option[FASTA]]` which returns `None` for the chunks that don't contain any valid FASTA header makes more sense. Then we still have 
```scala
def parseFastaDropErrors: Iterator[FASTA] = 
  parseFasta.flatten
```

And the other iterator I meantion above, which parses only consecutive valid FASTAs will be defined like this: 
```scala
def parseConsecutiveFasta: Iterator[FASTA] = 
  parseFasta.takeWhile(_.isDefined).flatten
```

laughedelic replied:
I just realised that currently once you got a line with a valid header, everything after it until the next valid header will be treated as the corresponding sequence: `cutUntil(Header.isValid)`, so most of the above doesn't make much sense. If there is no way to know where a sequence ends and some crap starts, there is no sense to return `Option`. We can only decide to drop or fail on some crap in the beginning (before the first valid header):

```scala
// implemented as it is now, but returns FASTA, not Option
def parseFasta: Iterator[FASTA] = {
  def hasNext: Boolean = lines.hasNext && Header.isValid(lines.head)
  // i.e. is empty if first line is not a valid header
}

// same, but skips crap until the first valid header
def parseFastaSkipCrap: Iterator[FASTA] = {
  cutUntil(Header.isValid)
  parseFasta
}
```

eparejatobes replied:
OK to https://github.com/ohnosequences/fastarious/pull/25/files#r144604543

laughedelic replied:
Done in 7ae79fa

laughedelic committed:
Space after id in the FASTA header is not a part of the description

laughedelic committed:
Added hasNext check to the FASTA iterator's next()

laughedelic commented:
@eparejatobes I made the changes you asked for. Also see https://github.com/ohnosequences/fastarious/pull/25#discussion_r143738636.

laughedelic committed:
Changed parseFasta to Iterator[FASTA]; added parseSkipCrap
Also removed parseFastaDropErrors (replaced with parseFasta).
If there is no way to know where a sequence ends and some crap starts,
there is no sense to return Option. We can only decide to drop or fail
on some crap in the beginning (before the first valid header).
